pkg/rtc/localparticipant.go

```diff
@@ -0,0 +1,375 @@
+	}
+
+	l.setupUptrackManager()
+
+	return l
+}
+
+func (l *LocalParticipant) Start() {
+	l.UptrackManager.Start()
+	go l.rtcpSendWorker()
+}
+
+func (l *LocalParticipant) Close() {
+	l.UptrackManager.Close()
+
+	l.pendingTracksLock.Lock()
+	l.pendingTracks = make(map[string]*pendingTrackInfo)
+	l.pendingTracksLock.Unlock()
+}
+
+func (l *LocalParticipant) OnWriteRTCP(f func(pkts []rtcp.Packet)) {
+	l.onWriteRTCP = f
+}
+
+func (l *LocalParticipant) OnTrackPublished(f func(track types.PublishedTrack)) {
+	l.onTrackPublished = f
+}
+
+// AddTrack is called when client intends to publish track.
+// records track details and lets client know it's ok to proceed
+func (l *LocalParticipant) AddTrack(req *livekit.AddTrackRequest) *livekit.TrackInfo {
+	l.pendingTracksLock.Lock()
+	defer l.pendingTracksLock.Unlock()
+
+	// if track is already published, reject
+	if l.pendingTracks[req.Cid] != nil {
+		return nil
+	}
+
+	if l.UptrackManager.GetPublishedTrackBySignalCidOrSdpCid(req.Cid) != nil {
+		return nil
+	}
+
+	ti := &livekit.TrackInfo{
+		Type:       req.Type,
+		Name:       req.Name,
+		Sid:        utils.NewGuid(utils.TrackPrefix),
+		Width:      req.Width,
+		Height:     req.Height,
+		Muted:      req.Muted,
+		DisableDtx: req.DisableDtx,
+		Source:     req.Source,
+		Layers:     req.Layers,
+	}
+	l.pendingTracks[req.Cid] = &pendingTrackInfo{TrackInfo: ti}
+
+	return ti
+}
+
+func (l *LocalParticipant) AddMigratedTrack(cid string, ti *livekit.TrackInfo) {
+	l.pendingTracksLock.Lock()
+	defer l.pendingTracksLock.Unlock()
+
+	l.pendingTracks[cid] = &pendingTrackInfo{ti, true}
+}
+
+func (l *LocalParticipant) SetTrackMuted(trackID livekit.TrackID, muted bool) {
+	track := l.UptrackManager.SetTrackMuted(trackID, muted)
+	if track != nil {
+		// handled in UptrackManager for a published track, no need to update state of pending track
+		return
+	}
+
+	isPending := false
+	l.pendingTracksLock.RLock()
+	for _, ti := range l.pendingTracks {
+		if livekit.TrackID(ti.Sid) == trackID {
+			ti.Muted = muted
+			isPending = true
+			break
+		}
+	}
+	l.pendingTracksLock.RUnlock()
+
+	if !isPending {
+		l.params.Logger.Warnw("could not locate track", nil, "track", trackID)
+	}
+}
+
+func (l *LocalParticipant) GetConnectionQuality() (scores float64, numTracks int) {
+	for _, pt := range l.UptrackManager.GetPublishedTracks() {
+		if pt.IsMuted() {
+			continue
+		}
+		scores += pt.GetConnectionScore()
+		numTracks++
+	}
+
+	return
+}
+
+func (l *LocalParticipant) GetDTX() bool {
+	l.pendingTracksLock.RLock()
+	defer l.pendingTracksLock.RUnlock()
+
+	var trackInfo *livekit.TrackInfo
+	for _, ti := range l.pendingTracks {
+		if ti.Type == livekit.TrackType_AUDIO {
+			trackInfo = ti.TrackInfo
+			break
+		}
+	}
+
+	if trackInfo == nil {
+		return false
+	}
+
+	return !trackInfo.DisableDtx
+}
+
+func (l *LocalParticipant) MediaTrackReceived(track *webrtc.TrackRemote, rtpReceiver *webrtc.RTPReceiver, mid string) (types.PublishedTrack, bool) {
```

Comment:
@cnderrauber This was in `UptrackManager`. It was moved here.

You were passing in `*ParticipantImpl` and doing the mid in this function. I did not want to pass in the parent pointer. So, calculating the `mid` at the call site and passing it in.

Also, this module does not have `onTrackUpdated` callback. So, returning indication if track is new and the call site fires `onTrackUpdated` if necessary.